packages/contracts/contracts/Unirep.sol

```diff
@@ -121,27 +138,24 @@ contract Unirep is IUnirep, VerifySignature {
         uint256[] calldata publicSignals,
         uint256[8] calldata proof
     ) public {
-        uint256 attesterId = publicSignals[2];
+        SignupSignals memory signals = decodeSignupSignals(publicSignals);
         // only allow attester to sign up users
-        if (uint256(uint160(msg.sender)) != attesterId)
-            revert AttesterIdNotMatch(uint160(msg.sender));
+        if (signals.dataHash != defaultUserDataHash) revert InvalidDataHash();
```

Comment:
I think this proof shouldn't include a `dataHash` value. Instead it can output the `stateLeafIdentityHash`. This function then can be overloaded like

```solidity
function userSignUp(
  uint[] calldata publicSignals,
  uint[8] calldata proof
) {}

function userSignUp(
  uint[] calldata publicSignals,
  uint[8] calldata proof,
  uint[] calldata data
) {}
```

If there is no data we calculate the state leaf as `H(stateLeafIdentityHash, defaultUserDataHash)`. If there is data we do it as `H(stateLeafIdentityHash, H(data))`.

Only the `stateLeafIdentityHash` _must_ be calculated in ZK because it has a secret value.